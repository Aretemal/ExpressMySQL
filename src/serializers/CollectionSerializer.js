import fullUrlCreator from '../utils/fullUrlCreator.js';
import Serializer from './Serializer.js';

class CollectionSerializer extends Serializer {
  constructor(resource, serializerType, request = null, metaData = null) {
    super(resource, request);
    this.serializerType = serializerType;
    this.metaData = metaData;
  }

  meta() {
    return this.metaData;
  }

  serialize() {
    const data = {
      data: this.collect(),
      links: this.links(),
    };
    if (this.metaData) {
      data.meta = this.meta();
    }
    return data;
  }

  collect() {
    return this.resource.map((item) => {
      const serializer = new this.serializerType(item);
      return serializer.serialize().data;
    });
  }

  links() {
    return { self: `${process.env.API_URL}/${this.request.originalUrl}` };
  }
}
export default CollectionSerializer;
