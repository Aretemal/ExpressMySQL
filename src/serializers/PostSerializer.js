import Serializer from './Serializer.js';

class PostSerializer extends Serializer {
  type() {
    return 'Post';
  }

  attributes() {
    return {
      authorId: this.resource.authorId,
      content: this.resource.content,
      createdAt: this.resource.createdAt,
    };
  }

  links() {
    return { self: `${process.env.API_URL}/post/${this.resource.id}` };
  }
}
export default PostSerializer;
