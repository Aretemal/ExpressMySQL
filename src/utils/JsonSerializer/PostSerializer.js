import Serializer from './Serializer.js';

class PostSerializer extends Serializer {
  create(data, links, id) {
    return this.serializer(data, id, 'Post', links);
  }

  getAll(data, links, id) {
    return this.serializer(data, id, 'Array Posts', links);
  }

  getOne(data, links, id) {
    return this.serializer(data, id, 'Post', links);
  }
}
export default new PostSerializer();
