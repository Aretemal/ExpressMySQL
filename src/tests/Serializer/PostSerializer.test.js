import { matchersWithOptions } from 'jest-json-schema';
import CollectionSerializer from '../../serializers/CollectionSerializer.js';
import PostSerializer from '../../serializers/PostSerializer.js';
import schema from './schema.json';

expect.extend(matchersWithOptions({
  verbose: true,
}));

describe('PostSerializer :', () => {
  describe('create :', () => {
    test('should return post', () => {
      const post = {
        id: 3,
        authorId: 2,
        content: 'Hello',
      };

      const serializer = new PostSerializer(post);
      const data = serializer.serialize();

      expect(data).toMatchSnapshot();
      expect(data).toMatchSchema(schema);
    });
  });

  describe('getAll :', () => {
    test('should return array posts', () => {
      const posts = [{
        id: 2,
        authorId: 2,
        content: 'Hello',
      },
      {
        id: 1,
        authorId: 1,
        content: 'World',
      }];
      const serializer = new CollectionSerializer(posts, { serializerType: PostSerializer });
      const data = serializer.serialize({ originalUrl: 'example' });

      expect(data).toMatchSnapshot();
      expect(data).toMatchSchema(schema);
    });
  });

  describe('getOne :', () => {
    test('should return post', () => {
      const post = {
        id: 1,
        authorId: 2,
        content: 'Hello',
      };
      const serializer = new PostSerializer(post);
      const data = serializer.serialize();

      expect(data).toMatchSnapshot();
      expect(data).toMatchSchema(schema);
    });
  });
});
