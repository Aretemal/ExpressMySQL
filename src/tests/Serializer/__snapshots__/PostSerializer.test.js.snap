// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostSerializer : create : should return post 1`] = `
{
  "data": {
    "attributes": {
      "authorId": 2,
      "content": "Hello",
    },
    "id": 3,
    "type": "Post",
  },
  "links": {
    "self": "http://localhost:5000/api/profile/posts",
  },
}
`;

exports[`PostSerializer : getAll : should return array posts 1`] = `
{
  "data": {
    "attributes": {
      "posts": [
        {
          "data": {
            "attributes": {
              "authorId": 2,
              "content": "Hello",
            },
            "id": 2,
            "type": "Post",
          },
          "links": {
            "self": "http://localhost:5000/api/post/2",
          },
        },
        {
          "data": {
            "attributes": {
              "authorId": 1,
              "content": "World",
            },
            "id": 2,
            "type": "Post",
          },
          "links": {
            "self": "http://localhost:5000/api/post/2",
          },
        },
      ],
    },
    "id": 0,
    "type": "Posts",
  },
  "links": {
    "self": "http://localhost:5000/api/profile/posts",
  },
}
`;

exports[`PostSerializer : getOne : should return post 1`] = `
{
  "data": {
    "attributes": {
      "authorId": 2,
      "content": "Hello",
    },
    "id": 1,
    "type": "Post",
  },
  "links": {
    "self": "http://localhost:5000/api/profile/posts/2",
  },
}
`;
